Automattic/jetpack
# Monorepo: Reorganize all projects

The Jetpack project has evolved over the years from a single product—Jetpack the plugin—to an ecosystem. The Jetpack repo is now the canonical home for block development and Composer packages, which may or may not ship in Jetpack the plugin.

Additionally, as Jetpack looks to offer more single-purpose or special-purpose plugins, the development requirements of setting up development infra in multiple places, to maintain them, to develop pieces of these products in various places will add a burden and slow development of those products.

Jetpack, as a monorepo, does not have a structure that is conducive to the successful development of other plugins—Jetpack is sprinkled in a bit of everywhere.

#### Changes proposed in this Pull Request:
* move Jetpack the plugin into a folder within a new `projects/plugins` directory.
* move all packages to `projects/packages`.
* move all `bin` and `docker` files to within the `tools` directory.
* split the root-level `package.json` and `composer.json` into a monorepo-specific one and a `projects/plugins/jetpack` one.
* adds a `jetpack.php` to gracefully fail for any sites that are running Jetpack directly checked-out of GitHub.
* improves the mirroring of all packages by adding an optional build step for projects that have JS (`lazy-images` is alone in this so far).
* mirrors Jetpack the plugin to a new https://github.com/Automattic/jetpack-production repo. The goal of this repo will be to have a virtually identical copy of Jetpack as what lands in SVN and ships. (Incomplete as part of this PR, but will be worked on this month)
* moves the `debug-helper` to the `projects/plugins` folder as it is a plugin, not a Composer package.


#### Jetpack product discussion
PT: p1HpG7-aSm-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Try to use the beta plugin to check out this branch. Confirm you get a functional Jetpack plugin.
 
Each of the other items has been specifically tested in PRs that merged into this feature branch:
* Confirm jetpack-production has Jetpack.
* Confirm the package mirror repos still have the mirrors.
* Confirm that Docker works (may need to `yarn docker:clean` or at least `yarn docker:build-image` before running `yarn docker:up`).
* May need to run `yarn clean`, then `yarn install` and `composer install` to freshen up the root.

#### Proposed changelog entry for your changes:
* No public facing changes.

#### Deployment notes:
* Need to ship D54963-code just before it (Updating Fusion).
